### ISSUE TITLE
MLflow + Ultralytics: Experiment logs are forcibly saved to /Shared/Ultralytics despite custom EXPERIMENT_NAME in mlflow.set_experiment(EXPERIMENT_NAME)

### ISSUE THREAD

**Issue by Ultraluxe25** [mlflow/mlflow#24130]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Summary

I was fine-tuning a YOLO26x-sem model and decided to track my experiment in the "Experiment 2" folder. I ran the code below in a Jupyter Notebook cell:

```python
%%time
settings.update({"mlflow": True})
DB_PATH = "my_experiment.db"
TRACKING_URI = f"sqlite:///my_experiment.db"

os.environ["MLFLOW_TRACKING_URI"] = TRACKING_URI
EXPERIMENT_NAME = "Experiment 2"

mlflow.set_tracking_uri(TRACKING_URI)
mlflow.set_experiment(EXPERIMENT_NAME)

print(f"✅ Tracking URI: {mlflow.get_tracking_uri()}")
print(f"✅ Database: {DB_PATH}")
print(f"✅ Experiment: {EXPERIMENT_NAME}")

exp = mlflow.get_experiment_by_name(EXPERIMENT_NAME)
print(f"✅ Experiment ID: {exp.experiment_id}")
```

and I received the following output:

```text
2026/06/22 21:18:29 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/06/22 21:18:30 INFO mlflow.store.db.utils: Updating database tables
2026/06/22 21:18:31 INFO mlflow.tracking.fluent: Experiment with name 'Experiment 2' does not exist. Creating a new experiment.
✅ Tracking URI: sqlite:///my_experiment.db
✅ Database: my_experiment.db
✅ Experiment: Experiment 2
✅ Experiment ID: 1
CPU times: total: 1.14 s
Wall time: 1.65 s
```

In the terminal, I started the MLflow server with:

```bash
mlflow ui --backend-store-uri sqlite:///my_experiment.db --host 0.0.0.0 --port 5000` command.
```

When I finished, I suddenly realized that all models were stored in the **Shared/Ultralytics folder**.

<img width="960" height="225" alt="Image" src="https://github.com/user-attachments/assets/9ed29e84-8c22-4eb9-85a3-e2b203120ff6" />

<img width="752" height="239" alt="Image" src="https://github.com/user-attachments/assets/d8293d34-b2ef-4dbd-a703-7227bca4c33d" />

and my **Experiment 2** folder was empty:

<img width="852" height="462" alt="Image" src="https://github.com/user-attachments/assets/c541179a-ae0a-45d0-be54-4cf4eaff295c" />

I tried placing **`settings.update({"mlflow": True})`** in different parts of the code cell at the beginning, but the result was still the same. I also read the MLflow documentation: https://docs.ultralytics.com/integrations/mlflow#introduction

Could you please explain why I received this result and how to correctly control the folder hierarchy in MLflow experiments in the future?

### Notes

I used Jupyter Notebook with YOLO CLI.
My MLFlow version is 3.14

**Comment by Rohityadav2005** [comment]
You are getting this issue because your ultralytics loggings overrides your manual loggings. So to solve this either you can disable your ultralytics builtin loggings or you can pass the experiment name directly to the training function.

**Comment by Charuhasini30** [comment]
Hi, I'd like to work on this issue.

I'll reproduce the behavior locally and investigate whether the Ultralytics integration is overriding the user-defined experiment configured via mlflow.set_experiment().

If I identify the root cause, I'll submit a PR with tests and/or documentation updates.

**Comment by ARJ8102** [comment]
I was able to reproduce this behavior locally.

with the Environment:

* OS: Windows
* Python: 3.12.9
* MLflow: 3.14.0
* Ultralytics: 8.4.80

When I used only:

mlflow.set_experiment("Experiment 2")


before training, the Ultralytics run was logged under:

/Shared/Ultralytics

and `Experiment 2` had 0 runs.

I then tested two alternatives.

 Option 1: Set `MLFLOW_EXPERIMENT_NAME`

os.environ["MLFLOW_TRACKING_URI"] = "sqlite:///my_experiment_env_fix.db"
os.environ["MLFLOW_EXPERIMENT_NAME"] = "Experiment 2"

Result:

- Experiment 2 | runs=1
- Default | runs=0


Option 2: Pass `project` to `model.train()`

model.train(
    data="coco8.yaml",
    epochs=1,
    imgsz=64,
    batch=1,
    project="Experiment 2",
    name="repro_project_fix_run",
)

Result:

- Experiment 2 | runs=1
- Default | runs=0


So it seems Ultralytics' MLflow integration does not use the experiment previously set with `mlflow.set_experiment()` during training. Instead, the experiment name can be controlled by setting `MLFLOW_EXPERIMENT_NAME` before training or by passing `project` in `model.train()`.

**Comment by harshavardhan194** [comment]
Hi, I'd like to work on this. I've traced the root cause to `on_pretrain_routine_end` in `mlflow.py` — it unconditionally calls `mlflow.set_experiment()` using the env var or `trainer.args.project`, overriding any experiment the user set beforehand. I'll implement a fix that respects a pre-configured experiment while keeping existing `MLFLOW_EXPERIMENT_NAME` and `project=` behavior intact, and submit a PR with tests.

**Comment by harshavardhan194** [comment]
I've reproduced the issue locally and investigated the root cause. I'd like to implement a fix with tests if this issue is ready for contribution.

**Comment by nj-codes-24** [comment]
Hey! I'd love to work on this if it's still available. I can put together a PR to make the documentation a bit clearer.

From what I understand, the confusion comes from the fact that the SQLite database only stores the experiment metadata (metrics, params, etc.), while the actual model artifacts are still being saved to Ultralytics' default output directory because the artifact location isn't being overridden.

I think it would help to mention that users can pass `project="Experiment 2"` directly to `model.train()` to control where the model files are saved. Another option, if they're creating experiments manually in MLflow, is to use `mlflow.create_experiment(..., artifact_location="path/to/folder")` to explicitly set the artifact location.

I also noticed a couple of small things that might be worth adding:

- Using `sqlite:///` creates a relative database path. Using `sqlite:////absolute/path/to/db` avoids accidentally creating multiple databases when running the script from different directories.

- mlflow.set_tracking_uri()` is enough on its own, so the `os.environ` line can be removed.

Happy to put together a PR for these changes if that works!

**Comment by PattaraS** [comment]
Thanks everyone for digging in — walking back my earlier label while we scope this correctly.

Two things are tangled here:

1. **The experiment override** — `on_pretrain_routine_end` unconditionally calling `mlflow.set_experiment()` lives in the Ultralytics package (`ultralytics/utils/callbacks/mlflow.py`), not in MLflow. MLflow's `set_experiment()` behaves as documented; the override behavior is Ultralytics' to change, so that part is best filed at https://github.com/ultralytics/ultralytics.

2. **Artifact location / config** (per @nj-codes-24) — some of this is genuinely MLflow-side (e.g. `create_experiment(artifact_location=...)`, absolute `sqlite:////` paths), while `project=` on `model.train()` is again Ultralytics.

@nj-codes-24 — is there a concrete, Ultralytics-independent gap in the MLflow docs you think is worth a small standalone PR? If so, let's scope that narrowly. Otherwise the core issue here is upstream. Marking `needs author feedback` while we sort out whether there's an MLflow-side item.

**Comment by nj-codes-24** [comment]
Thanks for breaking that down! That makes a lot of sense, and I agree the core behavior override (and the `project=` argument) should be addressed upstream in Ultralytics.

To answer your question—yes, for the MLflow side, I think there are a few concrete, Ultralytics-independent documentation gaps we could fill to help prevent similar confusion for users setting up local tracking. 

I’d be happy to put together a narrowly-scoped PR to add the following:

* **Clarifying SQLite Pathing:** Add a brief note in the Tracking (backend store) docs highlighting the difference between relative (`sqlite:///`) and absolute (`sqlite:////`) paths. This is a common gotcha that leads users to accidentally create multiple disconnected databases when running scripts from different directories.
* **API vs. Environment Variable:** Briefly clarify in the tracking examples or `set_tracking_uri` docs that calling `mlflow.set_tracking_uri()` is fully sufficient. This ensures users know they don't need to manually set `os.environ["MLFLOW_TRACKING_URI"]` inside their Python scripts as well.
* **Local Artifact Locations:** Explicitly highlight in the local tracking examples that while the SQLite DB handles metadata, users can control where local artifacts are saved by passing the `artifact_location` parameter to `mlflow.create_experiment()`.

Let me know if this scope works for you, and I can get a PR started!
@PattaraS

**Comment by Ultraluxe25** [comment]
@ARJ8102 Thank you very much for your help, and thanks a lot to all the other advisers!

## Option 2: Pass project to model.train() helped!

```jupyter notebook
# jupyter notebook cell 1:
TRACKING_URI = "sqlite:///mlflow1.db"
settings.update({"mlflow": True})
print(f"✅ Tracking URI: {mlflow.get_tracking_uri()}")

# jupyter notebook cell 2:
%%time
!yolo task=semantic mode=train model=yolo26n-sem.pt data='data.yaml' batch=16 device=0 epochs=25 imgsz=640 plots=True project="Semantic segmentation" name="25 epochs" workers=0
```

**project="Semantic segmentation" name="25 epochs"**

and in terminal `mlflow ui --backend-store-uri sqlite:///mlflow1.db --port 5000`

<img width="565" height="232" alt="Image" src="https://github.com/user-attachments/assets/0f119649-cac3-4d17-a4ca-43b4f0c6b847" />
<img width="760" height="283" alt="Image" src="https://github.com/user-attachments/assets/dc43b64d-f78b-4d25-ae0a-e0b8587331a0" />
